### PR TITLE
Removing constant folding of `!`

### PR DESCRIPTION
**Pull request by aaronjeline**
## Description of changes
Removes the constant folding `!` in the CST->AST parser.
The motivation for this change is that the CST->AST parser does this simplification, but the EST->AST parser does not. We'd like these parsers to line up, so we either need to remove it from one or add it to the other. I'm voting for removing it for simplicity, having all these little simplifications throughout the parser bugs me. They are hard to audit/verify and make sure we've gotten them right, and hard to ensure consistency between our different parsers. (It's bitten us here and in the past)
## Issue #, if available
N/A
## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [X] A bug fix or other functionality change requiring a patch to `cedar-policy`.

I confirm that this PR (choose one, and delete the other options):

- [X] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [X] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.

### CODE REVIEW

File: cedar-policy-core/src/parser/cst_to_ast.rs
Comment (john-h-kastner-aws) [inline cedar-policy-core/src/parser/cst_to_ast.rs:1]: Have you looked at what we're doing for `-`?

**Review by john-h-kastner-aws — APPROVED** [review]
I like this change.

Some additional context: This difference was found by the `convert-policy-cedar-to-json` fuzz target. For the target to pass we need to have the EST and CST parsing paths produce the same AST, so either folding `!` or not. The transformation feels unnecessary, and I would prefer not to duplicate it and maintain it in two places, so removing it in the CST->AST function seems appropriate.

File: cedar-policy-core/src/parser/cst_to_ast.rs
Comment (aaronjeline) [reply cedar-policy-core/src/parser/cst_to_ast.rs:1]: Yeah, `-` doesn't fold

**Review by cdisselkoen — APPROVED** [review]
This makes me wish we had a generic AST->AST constant folding pass though

**Pull request merged**